File: typescript/hard/median-of-two-sorted-arrays.ts
//--- LEETCODE

function findMedianSortedArrays(nums1: number[], nums2: number[]): number {
  // Garante que nums1 seja o menor array
  if (nums1.length > nums2.length) {
    return findMedianSortedArrays(nums2, nums1);
  }

  const x = nums1.length;
  const y = nums2.length;

  let low = 0, high = x;

  while (low <= high) {
    const partitionX = Math.floor((low + high) / 2);
    const partitionY = Math.floor((x + y + 1) / 2) - partitionX;

    const maxLeftX = partitionX === 0 ? -Infinity : nums1[partitionX - 1];
    const minRightX = partitionX === x ? Infinity : nums1[partitionX];

    const maxLeftY = partitionY === 0 ? -Infinity : nums2[partitionY - 1];
    const minRightY = partitionY === y ? Infinity : nums2[partitionY];

    if (maxLeftX <= minRightY && maxLeftY <= minRightX) {
      // Verifica se o total de elementos é ímpar
      if ((x + y) % 2 === 1) {
        return Math.max(maxLeftX, maxLeftY);
      }
      // Total de elementos é par
      return (Math.max(maxLeftX, maxLeftY) + Math.min(minRightX, minRightY)) / 2;
    } else if (maxLeftX > minRightY) {
      high = partitionX - 1;
    } else {
      low = partitionX + 1;
    }
  }

  throw new Error("Input arrays are not sorted");
}

// Testes
const nums1 = [1, 3];
const nums2 = [2];
console.log(findMedianSortedArrays(nums1, nums2)); // Output: 2.0

const nums3 = [1, 2];
const nums4 = [3, 4];
console.log(findMedianSortedArrays(nums3, nums4)); // Output: 2.5

File: golang/hard/median-of-two-sorted-arrays.go
package main

import (
	"fmt"
	"math"
)

func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
	// Garante que nums1 seja o menor array
	if len(nums1) > len(nums2) {
		return findMedianSortedArrays(nums2, nums1)
	}

	x, y := len(nums1), len(nums2)
	low, high := 0, x

	for low <= high {
		partitionX := (low + high) / 2
		partitionY := (x+y+1)/2 - partitionX

		// Pega os elementos em volta da partição
		maxLeftX := math.MinInt32
		if partitionX > 0 {
			maxLeftX = nums1[partitionX-1]
		}

		minRightX := math.MaxInt32
		if partitionX < x {
			minRightX = nums1[partitionX]
		}

		maxLeftY := math.MinInt32
		if partitionY > 0 {
			maxLeftY = nums2[partitionY-1]
		}

		minRightY := math.MaxInt32
		if partitionY < y {
			minRightY = nums2[partitionY]
		}

		// Verifica se encontramos a partição correta
		if maxLeftX <= minRightY && maxLeftY <= minRightX {
			// Se o número total de elementos for ímpar
			if (x+y)%2 == 1 {
				return float64(max(maxLeftX, maxLeftY))
			}
			// Se for par, retorna a média dos dois elementos centrais
			return float64(max(maxLeftX, maxLeftY)+min(minRightX, minRightY)) / 2
		} else if maxLeftX > minRightY {
			// Move para a esquerda
			high = partitionX - 1
		} else {
			// Move para a direita
			low = partitionX + 1
		}
	}

	panic("Input arrays are not sorted")
}

// Função para pegar o maior número entre dois
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// Função para pegar o menor número entre dois
func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	// Testes
	nums1 := []int{1, 3}
	nums2 := []int{2}
	fmt.Println(findMedianSortedArrays(nums1, nums2)) // Output: 2.0

	nums1 = []int{1, 2}
	nums2 = []int{3, 4}
	fmt.Println(findMedianSortedArrays(nums1, nums2)) // Output: 2.5
}
